File: android/app/src/main/kotlin/com/myapp/cyber_guard/SecurityEventEmitter.kt
package com.myapp.cyber_guard

import android.os.Handler
import android.os.Looper
import io.flutter.plugin.common.EventChannel

/**
 * Emits security events from Android native code to Flutter Dart.
 *
 * ## How EventChannel works:
 * 1. Dart calls `EventChannel.receiveBroadcastStream()` to start listening
 * 2. Flutter calls `onListen()` on this handler — we get an `EventSink`
 * 3. We call `eventSink.success(data)` to send events to Dart
 * 4. Dart receives the data in the stream listener
 * 5. When Dart cancels the stream, `onCancel()` is called
 *
 * ## Why a dedicated emitter:
 * - Separates event emission from plugin logic
 * - Thread-safe: all sink calls happen on the main thread via Handler
 * - Multiple producers (monitoring thread, callbacks) feed into one sink
 */
class SecurityEventEmitter : EventChannel.StreamHandler {

    private var eventSink: EventChannel.EventSink? = null
    private val mainHandler = Handler(Looper.getMainLooper())

    /**
     * Called by Flutter when Dart starts listening to the EventChannel.
     * We store the sink reference for later use.
     */
    override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
        eventSink = events
    }

    /**
     * Called by Flutter when Dart stops listening.
     * We clear the sink to prevent sending to a closed stream.
     */
    override fun onCancel(arguments: Any?) {
        eventSink = null
    }

    /**
     * Send a security event to Dart.
     *
     * Thread-safety: This can be called from ANY thread (monitoring thread,
     * MediaProjection callback, broadcast receiver). We post to the main
     * thread because EventSink.success() must be called on the UI thread.
     *
     * @param type The SecurityEventType name (must match Dart enum exactly)
     * @param severity One of: "low", "medium", "high", "critical"
     * @param metadata Optional additional data for forensic logging
     */
    fun emit(
        type: String,
        severity: String = "high",
        metadata: Map<String, Any> = emptyMap()
    ) {
        mainHandler.post {
            eventSink?.success(
                mapOf(
                    "type" to type,
                    "severity" to severity,
                    "timestamp" to System.currentTimeMillis(),
                    "metadata" to metadata
                )
            )
        }
    }

    /** Whether anyone is currently listening. */
    val isListening: Boolean get() = eventSink != null
}

File: ios/Runner/Security/SecurityEventEmitter.swift
import Flutter
import Foundation

/// Emits security events from iOS native code to Flutter Dart.
///
/// ## How FlutterEventChannel works on iOS:
///
/// 1. Dart calls `EventChannel.receiveBroadcastStream()` to start listening
/// 2. Flutter calls `onListen()` on this handler — we get a `FlutterEventSink`
/// 3. We call `eventSink(data)` to send events to Dart
/// 4. When Dart cancels, `onCancel()` is called
///
/// ## Thread Safety:
/// `FlutterEventSink` must be called on the main thread.
/// All emit() calls dispatch to DispatchQueue.main.
class SecurityEventEmitter: NSObject, FlutterStreamHandler {

    /// The sink provided by Flutter when Dart starts listening.
    private var eventSink: FlutterEventSink?

    /// Whether anyone is currently listening.
    var isListening: Bool { eventSink != nil }

    // MARK: - FlutterStreamHandler

    /// Called when Dart starts listening to the EventChannel.
    func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? {
        self.eventSink = events
        return nil
    }

    /// Called when Dart stops listening.
    func onCancel(withArguments arguments: Any?) -> FlutterError? {
        self.eventSink = nil
        return nil
    }

    // MARK: - Event Emission

    /// Send a security event to Dart.
    ///
    /// Dispatches to main thread because FlutterEventSink must be called
    /// from the platform thread (main/UI thread on iOS).
    ///
    /// - Parameters:
    ///   - type: SecurityEventType name (must match Dart enum: "screenCapture", "debuggerAttached", etc.)
    ///   - severity: One of "low", "medium", "high", "critical"
    ///   - metadata: Optional additional data for forensic logging
    func emit(type: String, severity: String = "high", metadata: [String: Any] = [:]) {
        guard isListening else { return }
        DispatchQueue.main.async { [weak self] in
            guard let self = self, let sink = self.eventSink else { return }
            sink([
                "type": type,
                "severity": severity,
                "timestamp": Int(Date().timeIntervalSince1970 * 1000),
                "metadata": metadata
            ])
        }
    }
}
